main.py

```diff
@@ -3,6 +3,166 @@
+#         r = alpha * g_padded[row - 1][column - 1] + (1 - alpha) * g_padded[row][column - 1]
+#         p = alpha * g_padded[row + 1][column + 1] + (1 - alpha) * g_padded[row][column + 1]
+
+#     if (intensity >= r) and (intensity >= p):
+#         return False
+#     else:
+#         return True
+    
+def pixel_suppressed(g, intensity, r_angle, row, column):
+    r_row, r_col, p_row, p_col = 0, 0, 0, 0
+
+    if r_angle == 0:
+        r_row = row
+        r_col = column + 1
+        p_row = row
+        p_col = column - 1
+    elif r_angle == 45:
+        r_row = row - 1
+        r_col = column + 1
+        p_row = row + 1
+        p_col = column - 1
+    elif r_angle == 90:
+        r_row = row - 1
+        r_col = column
+        p_row = row + 1
+        p_col = column
+    else:
+        r_row = row - 1
+        r_col = column - 1
+        p_row = row + 1
+        p_col = column + 1
+
+    if 0 <= r_row < g.shape[0] and 0 <= r_col < g.shape[1]:
+        r = g[r_row][r_col]
+    else:
+        r = 0
+
+    if 0 <= p_row < g.shape[0] and 0 <= p_col < g.shape[1]:
+        p = g[p_row][p_col]
+    else:
+        p = 0
+
+    if (intensity >= r) and (intensity >= p):
+        return False
+    else:
+        return True
+
+def canny(image, t_low, t_high):
+    image_blur = cv2.GaussianBlur(image, (3, 3), 1)
+    edges = cv2.Canny(image_blur, 50, 200, None, 3)
+    #  Build Sobel filters
+    gx_mask = np.array([[-1, 0, 1],
+                        [-2, 0, 2],
+                        [-1, 0, 1]])
+    
+    gy_mask = np.array([[1, 2, 1],
+                        [0, 0, 0],
+                        [-1, -2, -1]])
+    
+    # Apply Sobel filters to image
+    # gx = cv2.filter2D(image_blur, ddepth=0, kernel=gx_mask)
+    # gy = cv2.filter2D(image_blur, ddepth=0, kernel=gy_mask)
+    gx = cv2.Sobel(image_blur, 0, 1, 0, ksize=3)
+    gy = cv2.Sobel(image_blur, 0, 0, 1, ksize=3)
+
+    # Calculate gradient magnitude and direction
+    g = np.hypot(gx, gy)
+    theta = np.arctan2(gy, gx)
+    g = (g / g.max()) * 255
+
+    # Convert to degrees
+    theta = np.rad2deg(theta)
+
+    suppress = np.zeros_like(image)
+
+    # Non-maximum suppression
+    for row in range(0, g.shape[0]):
+        for column in range(0, g.shape[1]):
+            angle = theta[row][column]
+            intensity = g[row][column]
+            r_angle = round_angle(angle)
+
+            if pixel_suppressed(g, intensity, angle, row + 1, column + 1) == False:
+                suppress[row][column] = g[row][column]
+
+    output = np.zeros_like(suppress)
+
+    output[suppress < t_low] = 0
+    output[suppress >= t_low] = t_low
+    output[suppress >= t_high] = 255
+
+    # Hysteresis Thresholding
+    for row in range(0, g.shape[0]):
+        for column in range(0, g.shape[1]):
+            if output[row][column] == 255:
+                new_r = row
+                new_c = column
+                intensity = output[row][column]
+                while(intensity >= t_low):
```

Comment:
This loop only traverses the edge in one direction. I would have thought you would need to traverse in both directions. I tried it myself and it didn't seem to doing anything. So not sure 🤷